### ISSUE TITLE
AWSIoT -[MQTTEncoder .cxx_destruct] random crashes

### ISSUE THREAD

**Issue by marcmatta** [aws/aws-sdk-ios#752]
* **What service are you using?**
AWSIoT

* **In what version of SDK are you facing the problem?** 
The issue started happening when we upgraded to v2.6.1. It was not happening on version 2.5.7.

* **Is the issue limited to Simulators / Actual Devices?** 
Actual Devices

* **Can your problem be resolved if you bump to a higher version of SDK?** 
Can't say as no similar issues were reported and fixed from 2.6.1 to 2.6.4. I have upgraded anyway and I'll update this ticket if other crashes are reported from the newest version.

* **Is this problem related to specific iOS version?**
Operating systems affected by this crash:
iOS 11 - 63%
iOS 10 - 35%
iOS 9 - 2%

* **How are you consuming the SDK? CocoaPods / Carthage / Prebuilt frameworks?**
CocoaPods

* **Can you give us steps to reproduce with a minimal, complete, and verifiable example? Please include any specific network conditions that might be required to reproduce the problem.**
Can't reproduce my self. I have attached the stack trace if that makes sense.
[crash mqttencoder .cxx_destruct.txt](https://github.com/aws/aws-sdk-ios/files/1410810/crash.mqttencoder.cxx_destruct.txt)

**Comment by marcmatta** [comment]
@carsonmcdonald could you please help me with this issue?

**Comment by bxpan** [comment]
Hi marcmatta,

Do you know when the crash occurred? What was the app doing at the time of crash? Or is it happening randomly?

Thanks.

**Comment by marcmatta** [comment]
Hi @bxpan,

I have the following detail from Crashlytics:

> BUG IN CLIENT OF LIBDISPATCH: Semaphore object deallocated while in use

The application was publishing Data using an AWSIoTDataManager instance. The stack trace that I have attached previously shows the following before the application crashed. 

> -[MQTTEncoder encodeMessage:]

**Comment by bxpan** [comment]
Hi Marcmatta,

How many AWSIotDataManager instance do you have ? Could you attach the logs after enabling verbose logs? 

Thanks.

**Comment by marcmatta** [comment]
Hi @bxpan ,

Sorry I was sick for the past week and I haven't checked this since. 

Anyway, below is how i am instantiating the dataManager instance:

```
public class IoT {
static var __once: () = {
        let configuration = AWSServiceConfiguration(
            region: .USWest2,
            credentialsProvider: credentialProvider)
        
        guard let theConfiguration = configuration else {
            return
        }
        
        AWSIoTData.register(with: theConfiguration, forKey: dataManagerKey)
        AWSIoTDataManager.register(with: theConfiguration, forKey: dataManagerKey)
}()
    
fileprivate lazy var iotDataManager: AWSIoTDataManager = {
      // dispatched once because from AWS documentation, calling this method twice can have unspecified behavior
        _ = KMFIoT.__once
        
        return AWSIoTDataManager(forKey: KMFIoT.dataManagerKey)
 }()
}

// This is a singleton instance
public class MyMainClass {
        public let iot = IoT()
        public static sharedInstance: MyMainClass = MyMainClass()
}
```

And then in other classes we use the iot when publishing data

**Comment by marcmatta** [comment]
Are you using KVO internally? because I've found the same issue with other libraries where they found that using KVO was the problem.

**Comment by marcmatta** [comment]
Can anyone help with this issue pls? We need to solve it as quickly as possible.

**Comment by bxpan** [comment]
Hi marcmatta,

Could you attach device log after enabling verbose logs? 

Thanks.

**Comment by keshavx11** [comment]
Is this issue resolved? 
if so, please share the resolution. I am facing the same crash.

**Comment by keshavx11** [comment]
**What service are you using?**
AWSIoT

**In what version of SDK are you facing the problem?**
v2.6.12

**Is the issue limited to Simulators / Actual Devices?**
Actual Devices

**Can your problem be resolved if you bump to a higher version of SDK?**
We are on the highest available version but still facing the crash.

**Is this problem related to specific iOS version?
Operating systems affected by this crash:**
iOS 11 - 86%
iOS 10 - 12%
iOS 9 - 2%

**How are you consuming the SDK? CocoaPods / Carthage / Prebuilt frameworks?**
CocoaPods

**Can you give us steps to reproduce with a minimal, complete, and verifiable example? Please include any specific network conditions that might be required to reproduce the problem.**
Can't reproduce myself. I have attached the stack trace if that makes sense.
[crash.mqttencoder.cxx_destruct.txt](https://github.com/aws/aws-sdk-ios/files/1726602/crash.mqttencoder.cxx_destruct.txt)

**Comment by kvasukib** [comment]
@marcmatta @keshavx11 Sorry for the delayed response.

The log messages https://github.com/aws/aws-sdk-ios/blob/master/AWSIoT/Internal/MQTTSDK/MQTTEncoder.m#L120,L130 and https://github.com/aws/aws-sdk-ios/blob/master/AWSIoT/Internal/MQTTSDK/MQTTEncoder.m#L180 will help us identify what happened in the encodeMessage. Do you have the verbose logs? If so, can you attach it here?

**Comment by ashtonmeuser** [comment]
This may not be the most helpful addition but here is the stack trace that I am seeing with this issue. I am also unable to reproduce the issue. Rather, it seems to happen randomly. 

Using AWS SDK 2.6.16, iOS 11.3 (15E216), iPhone 8 (Model A1863, A1905, A1906, A1907)

![Image](https://i.imgur.com/iZDXdNn.jpg)

**Comment by kvasukib** [comment]
Thank you @ashtonmeuser for the details. Would you be able to collect the logs from the XCode console and attach it here?

**Comment by keshavx11** [comment]
2018-02-23 14:57:14:136 Intouchapp[988:529545] Auto-resubscribe is enabled. Resubscribing to topics.
2018-02-23 14:57:14:137 Intouchapp[988:529545] MQTT session error, code: 3
2018-02-23 14:57:14:138 Intouchapp[988:529545] closing encoder stream.
2018-02-23 14:57:14:138 Intouchapp[988:529545] closing decoder stream.
2018-02-23 14:57:14.162416+0530 Intouchapp[988:536637] [] nw_socket_get_input_frames recvmsg(fd 19, 1024 bytes): [57] Socket is not connected
2018-02-23 14:57:14.162615+0530 Intouchapp[988:536637] [] nw_socket_get_input_frames recvmsg(fd 24, 1024 bytes): [57] Socket is not connected
30 Intouchapp[988:536637] [] nw_socket_get_input_frames recvmsg(fd 38, 1024 bytes): [57] Socket is not connected
2018-02-23 14:57:14.172167+0530 Intouchapp[988:536637] [] nw_endpoint_handler_add_write_request [82.1 216.58.203.168:443 failed socket-flow (satisfied)] cannot accept write requests
2018-02-23 14:57:14.172266+0530 Intouchapp[988:536637] [] nw_endpoint_handler_add_write_request [83.1 35.164.249.251:443 failed socket-flow (satisfied)] cannot accept write requests
1 54.235.111.150:443 failed socket-flow (satisfied)] cannot accept write requests
2018-02-23 14:57:14.180813+0530 Intouchapp[988:538735] [] tcp_connection_write_eof_block_invoke Write close callback received error: [22] Invalid argument
2018-02-23 14:57:14.180898+0530 Intouchapp[988:534933] [] tcp_connection_write_eof_block_invoke Write close callback received error: [22] Invalid argument
2018-02-23 14:57:14.187399+0530 Intouchapp[988:536637] [] nw_endpoint_handler_add_write_request [83.1 35.164.249.251:443 failed socket-flow (satisfied)] cannot accept write requests
2018-02-23 14:57:14.188530+0530 Intouchapp[988:534965] [] tcp_connection_write_eof_block_invoke Write close callback received error: [22] Invalid argument
2018-02-23 14:57:15:140 Intouchapp[988:529545] MQTT session connection error


These were the last logs I recorded when I unexpectedly got the crash in development mode.
Hope they are useful @kvasukib.

**Comment by ashtonmeuser** [comment]
@kvasukib I did not have verbose logging enabled for the SDK. I'll enable it and submit the logs if/when I encounter this error again.

**Comment by steelzeh** [comment]
Any information on this, i'm getting a crash randomly whenever the internet is disconnected, and i call iotDataManager.disconnect()

<img width="369" alt="screen shot 2018-04-25 at 10 46 51" src="https://user-images.githubusercontent.com/7465835/39235561-78dae6cc-4876-11e8-94ae-fd59fde95537.png">

**Comment by harryzhao1** [comment]
I am also experiencing this crash.
Whenever when the user goes to home screen and re-enters the app, we disconnect from IOT and then reconnect.
This crash occurs sometimes randomly. It seems like something is strongly referenced instead of weak in the IOT SDK.

**Comment by steelzeh** [comment]
Will this ever be fixed, been an issue since 24'th of October 2017.

**Comment by cbommas** [comment]
@steelzeh 

I am actively working on this issue and hope to release a fix fairly soon. I will make sure to loop back on this thread.

**Comment by bhabani-sixt** [comment]
Looks like the crashes are happening in main thread. Can we avoid connect/disconnect getting executed from main thread.

**Comment by keshavx11** [comment]
Any updates on this issue?

**Comment by aat2703** [comment]
@cbommas any news?

**Comment by cbommas** [comment]
hello all,

We are targeting a bunch of fixes to address these issues in the next rev of the SDK. 
I will circle back once the release is out.

**Comment by aat2703** [comment]
@cbommas sounds great... Any idea of a possible release date (+-) yet?

**Comment by cbommas** [comment]
hello @aat2703 

The next release is slated to go out by the end of the week. I will let you know once I get confirmation that the fixes made it into the release.

**Comment by aat2703** [comment]
Great @cbommas ! Our customers is experiencing a lot of interruptions because of crashes/reconnect issues

Have you found the core issue with the crashes yet?

I'll wait for your confirmation 👍

**Comment by cbommas** [comment]
hello @aat2703, all

The bulk of the crashes were around the disconnect logic which had a few bugs in it. We released version 2.6.19 today with bug fixes and architectural changes which should resolve the bulk of the crashes. Can you give it a try and let me know how it goes.
